Implement comments reporting

Feodor0090 opened the pull request:
Adds "report" button.
Request scheme was written off chrome devtools by sending reports on dev server.
Form design was implemented just to work, i did not even trу to follow web design. Can try to make it closer on explicit request.
![изображение](https://user-images.githubusercontent.com/53872073/195866247-8956d611-a332-4fcd-be29-588e0fadf7e2.png)

Feodor0090 committed:
Add basic UI for reporting

Feodor0090 committed:
Add request class

Feodor0090 committed:
Implement reporting flow

Feodor0090 committed:
Add test

Feodor0090 committed:
Merge branch 'master' into comment-report

frenzibyte commented on osu.Game/Overlays/Comments/ReportCommentPopover.cs:
```diff
@@ -0,0 +1,83 @@
+            string infoValue = info.Current.Value;
+            var reasonValue = reason.Current.Value;
+
+            if (reasonValue == CommentReportReason.Other && string.IsNullOrWhiteSpace(infoValue))
```
I'm not sure where you got this behaviour from, but that's not how it works in web. "Commnets" section is required to submit a report.

frenzibyte commented on osu.Game/Overlays/Comments/ReportCommentPopover.cs:
```diff
@@ -0,0 +1,83 @@
+                Spacing = new Vector2(7),
+                Children = new Drawable[]
+                {
+                    reason = new LabelledEnumDropdown<CommentReportReason>
+                    {
+                        Label = "Reason"
+                    },
+                    info = new LabelledTextBox
+                    {
+                        Label = "Additional info",
+                    },
```
Use localised web strings instead (same applies everywhere a string is used for display in this PR, when possible). See `UsersStrings.Report*`.

frenzibyte commented on osu.Game/Overlays/Comments/CommentReportReason.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace osu.Game.Overlays.Comments
+{
+    public enum CommentReportReason
+    {
+        Insults,
+        Spam,
+        UnwantedContent,
+        Nonsense,
+        Other
+    }
```
Add localised descriptions. See usages of `LocalisableDescriptionAttribute`.

frenzibyte commented on osu.Game/Online/API/Requests/CommentReportRequest.cs:
```diff
@@ -0,0 +1,39 @@
+            Info = info;
+        }
+
+        protected override WebRequest CreateWebRequest()
+        {
+            var req = base.CreateWebRequest();
+            req.Method = HttpMethod.Post;
+
+            req.AddParameter(@"reportable_type", @"comment");
+            req.AddParameter(@"reportable_id", $"{CommentID}");
+            req.AddParameter(@"reason", Reason.ToString());
+            if (!string.IsNullOrWhiteSpace(Info))
+                req.AddParameter(@"comments", Info);
+
+            return req;
+        }
+
+        protected override string Target => @"reports";
```
There is no existing documentation for the `reports` endpoint. Can you provide how you extracted the parameter names here, and how you found `comments` to be optional?

Feodor0090 replied:
Parameter names:
![изображение](https://user-images.githubusercontent.com/53872073/196047631-f5b17cd7-2f92-4f3a-9cba-b333880df7c9.png)
Comment is requered to submit, my mistake.

frenzibyte commented on osu.Game/Overlays/Comments/DrawableComment.cs:
```diff
@@ -544,5 +572,10 @@ private string getParentMessage()
             }
         }
+
+        public Popover GetPopover()
+        {
+            return new ReportCommentPopover(ReportComment);
+        }
```
Use lambda for single-line code like this.
```suggestion
        public Popover GetPopover() => new ReportCommentPopover(ReportComment);
```

frenzibyte commented on osu.Game/Overlays/Comments/DrawableComment.cs:
```diff
@@ -403,6 +409,28 @@ private void deleteCommentRequest()
+            request.Success += () => Schedule(() =>
+            {
+                actionsLoading.Hide();
+                notificationOverlay?.Post(new SimpleNotification
+                {
+                    Icon = FontAwesome.Solid.CheckCircle,
+                    Text = "The comment reported successfully."
+                });
```
Not sure this should exist, but an OSD similar to the one in `ExternalLinkButton` may be better than a full notification.

frenzibyte reviewed (changes requested):
Crashes when clicking "report" in beatmap info overlay.
```
[runtime] 2022-10-16 00:19:24 [verbose]: Unhandled exception has been denied .
Unhandled exception. System.InvalidOperationException: Cannot show or hide a popover without a parent PopoverContainer in the hierarchy
   at osu.Framework.Extensions.PopoverExtensions.setTargetOnNearestPopover(Drawable origin, IHasPopover target)
   at osu.Framework.Extensions.PopoverExtensions.ShowPopover(IHasPopover hasPopover)
   at osu.Game.Graphics.Containers.LinkFlowContainer.<>c__DisplayClass15_0.<createLink>b__0() in /Users/salman/Projects/osu/osu.Game/Graphics/Containers/LinkFlowContainer.cs:line 84
   at osu.Framework.Graphics.Containers.ClickableContainer.OnClick(ClickEvent e)
```

---

For the design, the popover looks too wide, and the dropdown has less space to contain long items (the popover also looks too spaced from the report button, which makes me not sure whether it's supposed to be a popover for the comment or a popover displayed on the centre of the screen). 

Generally, I don't think `Labelled*` components are a good choice for popovers, they're best suited for full screen areas like the editor setup screen where they're used.

I've tinkered around with this and I've came up with another `PopupDialog`-based design, but I can't say I'm a big fan of the ending result... (can pass code on request)

<img width="1707" alt="CleanShot 2022-10-16 at 03 38 46@2x" src="https://user-images.githubusercontent.com/22781491/196012568-1cc148e1-a737-42c7-a901-a0ad4ce552bf.png">

A custom implementation that matches web design might be best, rather than coming up with new ones.

Feodor0090 committed:
Add localization for enum

Feodor0090 committed:
Display toast instead of notification

Feodor0090 commented:
Made a form with web's design.
![изображение](https://user-images.githubusercontent.com/53872073/196048549-a8d32f90-4178-4a60-b385-da7edd4641d7.png)
If it's okay, will and animations and push, but i'm not sure how to display it. `DialogOverlay` is made for this, but accepts only `PopupDialog`s. Should i inherit it and just overwrite all the content?

Feodor0090 committed:
Make report's comment not optional

frenzibyte commented:
I think a good approach would be to define an `IPopupDialog` interface publicly used by `DialogOverlay` and implement that in your custom implementation. Not entirely sure if that would work but would be an amicable one that avoids rewriting existing implementation or adding new local logic for it.

Feodor0090 committed:
Make a new report form, closer to web

Feodor0090 commented:
WIP, waiting for #20796.

Feodor0090 committed:
Add popover containers to overlays

Feodor0090 committed:
Update popover not to use labelled drawables

Feodor0090 committed:
Delete wip form

Feodor0090 committed:
Update test

Feodor0090 commented:
Decided to use popover.

![изображение](https://user-images.githubusercontent.com/53872073/196159436-4ca0a37a-947f-4742-92a2-8159107de33e.png)

I updated it a little to look closer to web and resolved issues found during review. Could you check this again?

> can pass code on request

What did you use to create "Report Someone's comment?" string? "Report" string is from UsersStrings, i found it, but how to format username?

Feodor0090 committed:
Merge branch 'master' into comment-report

Feodor0090 committed:
Fix reason not set in test

Feodor0090 committed:
Show username in popup

Feodor0090 committed:
Localize the button

Feodor0090 committed:
Merge branch 'comment-report' of https://github.com/Feodor0090/osu into comment-report

peppy committed:
Merge branch 'master' into comment-report

Feodor0090 committed:
Fix accidental breakage of changelog layout

frenzibyte commented on osu.Game/Overlays/Comments/DrawableComment.cs:
```diff
@@ -31,7 +34,7 @@
 namespace osu.Game.Overlays.Comments
 {
-    public class DrawableComment : CompositeDrawable
+    public class DrawableComment : CompositeDrawable, IHasPopover
```
The comment shouldn't be the popover target, that makes the popover display in completely unexpected positions:

https://user-images.githubusercontent.com/22781491/196821619-742ddc33-54f2-459f-86ad-7d443b5529da.mp4

The "Report" text should be the target. That should be possible by moving the text to a separate (`ReportCommentLabel`) component that inherits `IHasPopover` and is added to the flow via `AddArbitraryDrawable`.

frenzibyte commented on osu.Game/Overlays/Comments/ReportCommentPopover.cs:
```diff
@@ -0,0 +1,111 @@
+                        Text = UsersStrings.ReportReason,
+                        Font = OsuFont.Torus.With(size: 20),
+                    },
+                    reason = new OsuEnumDropdown<CommentReportReason>
+                    {
+                        RelativeSizeAxes = Axes.X
+                    },
```
The dropdown shouldn't autosize the popover when opening it, that results in bad UX:

https://user-images.githubusercontent.com/22781491/196821976-532ce76a-d5eb-4890-bcb8-bcce4cd8879f.mp4

Should work similar to multiplayer match settings.

frenzibyte commented on osu.Game/Overlays/Comments/ReportCommentPopover.cs:
```diff
@@ -0,0 +1,111 @@
+            string infoValue = info.Current.Value;
+            var reasonValue = reason.Current.Value;
+
+            if (string.IsNullOrWhiteSpace(infoValue))
+            {
+                shaker.Shake();
+                return;
+            }
```
Not sure if it's only me, but I would prefer button displaying in disabled state rather than shaking.

frenzibyte commented on osu.Game/Overlays/Comments/ReportCommentPopover.cs:
```diff
@@ -0,0 +1,111 @@
+                    {
+                        Origin = Anchor.TopCentre,
+                        Anchor = Anchor.TopCentre,
+                        Text = UsersStrings.ReportTitle(comment.Comment.User?.Username ?? comment.Comment.LegacyName!),
```
```suggestion
                        Text = ReportStrings.CommentTitle(comment.Comment.User?.Username ?? comment.Comment.LegacyName!),
```

frenzibyte commented on osu.Game/Overlays/Comments/DrawableComment.cs:
```diff
@@ -560,5 +581,15 @@ private string getParentMessage()
+
+        public Popover GetPopover() => new ReportCommentPopover(this);
+
+        private class ReportToast : Toast
+        {
+            public ReportToast()
+                : base(UserInterfaceStrings.GeneralHeader, UsersStrings.ReportThanks, "")
+            {
+            }
+        }
```
Looking at this again, maybe text placed in place of the "report" button would feel better than a toast:

https://user-images.githubusercontent.com/22781491/196823678-ceec56b6-5a01-4b3c-bb0c-a77a47835994.mp4

@peppy thoughts?

frenzibyte replied:
The text proposal has already been pushed in this PR, but I'll wait for further opinions before merge.

frenzibyte commented on osu.Game/Overlays/Comments/DrawableComment.cs:
```diff
@@ -413,6 +416,24 @@ private void deleteCommentRequest()
 
+        public void ReportComment(CommentReportReason reason, string comment)
+        {
+            actionsContainer.Hide();
```
Not sure reporting a comment should hide all actions? One may want to take a permalink of the comment after reporting it, for example.

frenzibyte commented on osu.Game/Overlays/BeatmapSetOverlay.cs:
```diff
@@ -44,20 +45,25 @@ public BeatmapSetOverlay()
             CommentsSection comments;
 
-            Child = new FillFlowContainer
+            Child = new PopoverContainer
```
Why is this not included in `CommentsContainer` instead?

Feodor0090 replied:
When there is only one comment, there is not enough space for the popover:
![изображение](https://user-images.githubusercontent.com/53872073/196994993-6111a295-6ed3-4a2e-bcf6-d32b563233d2.png)
I was not sure that it will work correctly in such cases, so made the whole overlay the container.

Feodor0090 committed:
Use another string for title

Feodor0090 committed:
Disable button when there is no text

Feodor0090 committed:
Make dropdown not resize

Feodor0090 committed:
Make report button a separate component

Feodor0090 committed:
Change test

Feodor0090 committed:
Display text in buttons flow instead of toast

frenzibyte committed:
Remove dependency on `DrawableComment` from report popover and simplify logic

Allows for testing the button and popover in isolation.

frenzibyte committed:
Adjust comment report popover design

frenzibyte committed:
Encapsulate report logic inside button implementation

Avoids complicating the `DrawableComment` class, and allows for isolated
testability.

frenzibyte committed:
Add visual test case for report button

Makes it much easier to test button/popover design changes

frenzibyte commented:
I've pushed several commits to clean up the code and keep things self-contained, along with changes to the popover design. Feels quite polished now (save for the popover size compared to the comment, which can't be helped without resolving the [bigger issue at hand](https://github.com/ppy/osu/issues/17940) first).

frenzibyte committed:
Inline content creation in base add method

peppy commented on osu.Game/Overlays/Changelog/ChangelogContent.cs:
```diff
@@ -7,20 +7,29 @@
 namespace osu.Game.Overlays.Changelog
 {
-    public class ChangelogContent : FillFlowContainer
+    public class ChangelogContent : PopoverContainer
```
I'm really not sure about having to add these all over the place. Did we discuss having a global one which can be the fallback case for whenever a local one is not required? Is there an issue with doing that?

The reason I bring this up is that this weird interaction happens due to the local `PopOverContainer` being padding too close:

https://user-images.githubusercontent.com/191335/197678961-255a6418-e1ae-43ac-9316-8b035b89d739.mp4

Feodor0090 replied:
I also had a question "why popover container isn't global". I added game-wide one inside `TooltipContainer`. Nothing seems to have broken from this.

peppy replied:
Bit hesitant to have that change bundled with this PR. At very least we'll want to go through all other usages of popovers as a follow-up and remove any local implementations that are no longer necessary (ie. anything that isn't in a scroll container and relying on the popover being localised to better track scroll).

peppy reviewed (commented):
Code looks pretty great apart from the pointed out issue.

Feodor0090 committed:
Add a global popover container

Feodor0090 committed:
Remove local popover containers

peppy committed:
Merge branch 'master' into comment-report

peppy commented:
So the immediate problem with moving the popover container to such a high level is it has the wrong colouring...

peppy committed:
Move `PopoverContainer` into `OnlineOverlay` to ensure correct colours

peppy commented:
I've moved it to `OnlineOverlay`, I think this is okay for now.
